### PR TITLE
Version 3.0.5

### PR DESCRIPTION
**Pull request by laurentdavid**
Updated tensorflow to the latest 2.5.x and numpy to 1.9.5 following dependabot recommendations.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -51,10 +51,10 @@
     },
     install_requires=[
         'matplotlib>=3.0,<3.1',
-        'numpy>=1.19.2,<1.20',
+        'numpy==1.19.2,<1.20',
```
Comment (douglasbagnall) [inline setup.py]: This looks strange to me. If we're insisting on `numpy==1.19.2`, why do we need to also say `<1.20`?

Also, I don't know why we would be so specific about numpy as to not accept 1.19.x bugfixes.

Having said that, I am sure the versions need tweaking.

Comment (laurentdavid) [reply setup.py]: Hi @douglasbagnall, 

Thanks. This is definitely wrong and it is a typo (I was trying different combinations at some point). I will fix this.

Laurent

**Comment by douglasbagnall** [comment]
@laurentdavid thanks, though the commit message still says 2.5.x. Also I think it is best to leave it more open, with something like `==2.7.*` or `>=2.7.1,<=2.8.0`.

@ilyatregubov: the workflow needs a maintainer to approve it.

This highlights that `setup.py` and `requirements.txt` have been saying different things, with the latter being used for CI.

**Comment by laurentdavid** [comment]
Hi @douglasbagnall,

Thanks, you are right, I set it to 2.7.* (I did not want to use 2.8.0 as it does not work ref. numpy dependencies).
Amended the commit message too.

Ref. the other dependencies, should we make sure that requirements.txt  => setup.py ? (maybe in another, newly created issue ?).

Laurent

**Comment by douglasbagnall** [comment]
> Ref. the other dependencies, should we make sure that requirements.txt => setup.py ? (maybe in another, newly created issue ?).

Let's wait for others to have their say.

**Comment by ilyatregubov** [comment]
Hi all. Yeah - it is not good that setup.py and requirements.txt contradict each other. Would be good to have them on sync.

I haven't check which version of tensorflow we want to use, but looking at discussion:

`==2.7.*` - This includes 2.7.0
`>=2.7.1,<=2.8.0` This doesn't

So say we don't wan't 2.7.0 as 2.7.1 has more fixes - we should use the latter

I have approved checks - seems to be green

**Comment by laurentdavid** [comment]
Hi @ilyatregubov & @douglasbagnall,

Maybe we should put more constraint on the version number i.e. use == instead of >= or =<.
From the documentation of pip (https://peps.python.org/pep-0440/#version-specifiers), from what I understand is that if you set >=2.7.0 or ==2.7.*, it should automatically take the greatest version number compatible with this release.
I would err on the side of caution on numpy and tensorflow and just set them as == so we know we are always in sync between setup.py/requirements.txt and the github test workflow.

Laurent

**Comment by ilyatregubov** [comment]
Hi @laurentdavid ,

Commit message still mentions numpy upgrade, but there is not change in numpy version

**Comment by douglasbagnall** [comment]
Regarding setup.py vs requirements.txt, it got this way because I added requirements.txt for the CI, and left setup.py alone on the assumption it was the distillation of accumulated wisdom, fitting a common use pattern, etc. In development I initially used direct pip installs, and now in deployment we use requirements.txt.

According to the Python canon, the two files *are* supposed to be different. https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/ says of setup.py (referred to as install_requires):

> It is not considered best practice to use install_requires to pin dependencies to specific  versions, or to specify sub-dependencies (i.e. dependencies of your dependencies). This is overly-restrictive, and prevents the user from gaining the benefit of dependency upgrades.

and of requirements.txt:

> Whereas install_requires defines the dependencies for a single project, [Requirements Files](https://pip.pypa.io/en/latest/user_guide/#requirements-files) are often used to define the requirements for a complete Python environment.

> Whereas install_requires requirements are minimal, requirements files often contain an exhaustive listing of pinned versions for the purpose of achieving [repeatable installations](https://pip.pypa.io/en/latest/user_guide/#repeatability) of a complete environment.

This is the reverse of what we do: we generally have stricter pinning in setup.py than in requirements.txt.

**Comment by laurentdavid** [comment]
Hi @douglasbagnall 

I see what you mean. Should we open a new issue on this topic ? I really wanted this pull request to be just about the tensorflow version change. I agree that there is room for improvement and happy to do it (being guided though as I am not very experienced on the dependencies topic in python).
I changed the commit message and relaxed the setup.py conditions related to tensorflow.

Laurent

**Comment by douglasbagnall** [comment]
Don't worry, @laurentdavid, I'm not wanting to torpedo your PR by conflating it with wider issues. Your commit should go in.

It's just I'm *also* using this thread to discuss the wider picture with Ilya and whoever else is watching.

**Comment by laurentdavid** [comment]
Hi @douglasbagnall,

Thanks for your understanding ! No problem and this is a useful discussion (as my first issue with setting up the project was that there was a discrepancy between the two, so a failure in the tests).
I created an issue here #56 so we can carry on with the discussion here and keep an eye on the changes ?
What do you think ?

Laurent

**Comment by ilyatregubov** [comment]
Agree on what is said here - if there is a flaw in release process - lets create an issue for it. Thanks for creating a separate issue for requirements.txt and setup.py handling.

As for the current issue for tensorflow upgrade - all tests are passing (pytests, Moodle unit tests, Moodle behat tests). And I also run few models manually to check the accuracy. So seems that this merge is ready to go.

**Pull request merged**